File: android/app/src/main/java/com/caganhatapci/orbeon/services/CustomSoundStore.kt
package com.caganhatapci.orbeon.services

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.ContextCompat
import android.os.Handler
import android.os.Looper
import com.caganhatapci.orbeon.store.prefs
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.File
import kotlin.concurrent.thread
import kotlin.math.abs
import kotlin.math.min

/**
 * Premium oyuncunun kendi sesiyle değiştirebildiği efekt yuvaları.
 * Atlayış sesi ayrı ele alınır: kayıt, kombo yükseldikçe hızlandırılıp
 * inceltilir — sentetik pentatonik dizinin yaptığını oyuncunun kendi sesi yapar.
 */
enum class CustomSoundSlot(val key: String, val maxMs: Int) {
    HOP("hop", 900),
    COLLECT("collect", 900),
    /**
     * Kapı açılma sesi. Bölüm bitirme sesinden AYRI bir yuva: ikisi aynıyken
     * oyuncu son yıldızı toplayınca bölümü bitirdiğini sanıyordu.
     */
    GATE("gate", 2000),
    LIFE_LOST("life_lost", 2000),
    FAIL("fail", 2000),
    WIN("win", 2000)
}

/**
 * Kayıtları cihazda tutar ve ses motoruna hazır PCM olarak verir.
 * Kayıtlar hiçbir yere yüklenmez.
 */
class CustomSoundStore(private val context: Context) {
    private val sampleRate = 44100

    /** Kaydı olan yuvalar */
    var recorded by mutableStateOf<Set<CustomSoundSlot>>(emptySet())
        private set
    /** Şu an kayıt alınan yuva */
    var recording by mutableStateOf<CustomSoundSlot?>(null)
        private set
    /**
     * Geri sayımı süren yuva. Mikrofona basınca kayıt HEMEN başlamıyor:
     * üçten geri sayılıyor. Aksi hâlde kaydın ilk yarısı, telefonu ağzına
     * götürürkenki hışırtı oluyordu.
     */
    var arming by mutableStateOf<CustomSoundSlot?>(null)
        private set
    /** Geri sayımda kalan saniye (3 → 2 → 1); kayıt dışında 0 */
    var countdown by mutableIntStateOf(0)
        private set
    /** Kayıtlar oyunda kullanılsın mı */
    var enabled by mutableStateOf(prefs(context).getBoolean(KEY_ENABLED, true))
        private set

    /** Faturalandırmadan beslenir; premium bitince kayıtlar devre dışı kalır */
    var premiumActive = false

    private var recordJob: Thread? = null
    /** Geri sayım adımları; vazgeçilirse hepsi birden iptal edilir */
    private val armHandler = Handler(Looper.getMainLooper())
    @Volatile private var stopRequested = false
    private var audio: AudioEngine? = null

    private val folder: File
        get() = File(context.filesDir, "customsfx").apply { if (!exists()) mkdirs() }

    private fun file(slot: CustomSoundSlot) = File(folder, "${slot.key}.pcm")

    init { refreshRecorded() }

    private fun refreshRecorded() {
        recorded = CustomSoundSlot.entries.filter { file(it).exists() }.toSet()
    }

    fun hasRecording(slot: CustomSoundSlot) = file(slot).exists()

    fun updateEnabled(value: Boolean) {
        enabled = value
        prefs(context).edit().putBoolean(KEY_ENABLED, value).apply()
        audio?.let { applyToEngine(it) }
    }

    fun hasMicPermission(): Boolean =
        ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) ==
            PackageManager.PERMISSION_GRANTED

    // MARK: Kayıt

    /**
     * 3 → 2 → 1 → kayıt. Her adımda kısa bir tık çalar; ekrana bakmadan da
     * ne zaman başlayacağı belli olsun diye.
     */
    fun startRecording(slot: CustomSoundSlot) {
        if (recording != null || arming != null || !hasMicPermission()) return
        arming = slot
        countdown = 3
        // İlk tıkı çağıran ekran çaldı; buradan bir daha çalmak çift ses olur
        for (step in 1..3) {
            armHandler.postDelayed({
                if (arming != slot) return@postDelayed
                val left = 3 - step
                countdown = left
                if (left > 0) {
                    audio?.playTap()
                } else {
                    arming = null
                    beginRecording(slot)
                }
            }, TICK_MS * step)
        }
    }

    /** Geri sayarken vazgeçildi */
    fun cancelArming() {
        armHandler.removeCallbacksAndMessages(null)
        arming = null
        countdown = 0
    }

    /**
     * Ham PCM yakalar. Süre dolunca kendiliğinden durur — uzun kayıt zaten
     * atlayış sesi olarak işe yaramıyor.
     */
    private fun beginRecording(slot: CustomSoundSlot) {
        countdown = 0
        val minBuf = AudioRecord.getMinBufferSize(
            sampleRate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT
        )
        if (minBuf <= 0) return

        val maxSamples = sampleRate * slot.maxMs / 1000
        stopRequested = false
        recording = slot

        recordJob = thread(start = true) {
            val rec = runCatching {
                @Suppress("MissingPermission")
                AudioRecord(
                    MediaRecorder.AudioSource.MIC, sampleRate,
                    AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT,
                    minBuf * 2
                )
            }.getOrNull()
            if (rec == null || rec.state != AudioRecord.STATE_INITIALIZED) {
                runCatching { rec?.release() }
                onMain { recording = null }
                return@thread
            }
            val out = ShortArray(maxSamples)
            var written = 0
            runCatching {
                rec.startRecording()
                val chunk = ShortArray(minBuf)
                while (!stopRequested && written < maxSamples) {
                    val n = rec.read(chunk, 0, min(chunk.size, maxSamples - written))
                    if (n <= 0) break
                    System.arraycopy(chunk, 0, out, written, n)
                    written += n
                }
            }
            runCatching { rec.stop() }
            runCatching { rec.release() }

            val clean = clean(out.copyOf(written))
            if (clean != null) save(slot, clean) else file(slot).delete()
            onMain {
                recording = null
                refreshRecorded()
                audio?.let { applyToEngine(it) }
            }
        }
    }

    /** Compose durumu yalnızca ana iş parçacığından yazılsın */
    private fun onMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) block()
        else Handler(Looper.getMainLooper()).post(block)
    }

    fun stopRecording() {
        // Geri sayarken "durdur"a basılırsa kayıt hiç başlamasın
        cancelArming()
        stopRequested = true
        recordJob = null
    }

    fun delete(slot: CustomSoundSlot) {
        file(slot).delete()
        refreshRecorded()
        audio?.let { applyToEngine(it) }
    }

    fun preview(slot: CustomSoundSlot, engine: AudioEngine) {
        load(slot)?.let { engine.playPreview(it) }
    }

    // MARK: Motora aktarma

    /**
     * Kayıtları (premium ve açıksa) motora yükler. Premium biterse motordaki
     * tamponlar boşaltılır ama dosyalar diskte kalır.
     */
    fun applyToEngine(engine: AudioEngine) {
        audio = engine
        if (!premiumActive || !enabled) {
            engine.clearCustomSounds()
            return
        }
        engine.applyCustomSounds(
            hop = load(CustomSoundSlot.HOP)?.let { engine.pitchLadder(it) } ?: emptyList(),
            collect = load(CustomSoundSlot.COLLECT),
            gate = load(CustomSoundSlot.GATE),
            lifeLost = load(CustomSoundSlot.LIFE_LOST),
            fail = load(CustomSoundSlot.FAIL),
            win = load(CustomSoundSlot.WIN)
        )
    }

    // MARK: Dosya + temizlik

    private fun save(slot: CustomSoundSlot, data: ShortArray) {
        runCatching {
            DataOutputStream(file(slot).outputStream().buffered()).use { out ->
                out.writeInt(data.size)
                data.forEach { out.writeShort(it.toInt()) }
            }
        }
    }

    private fun load(slot: CustomSoundSlot): ShortArray? {
        val f = file(slot)
        if (!f.exists()) return null
        return runCatching {
            DataInputStream(f.inputStream().buffered()).use { input ->
                val n = input.readInt()
                if (n <= 0 || n > sampleRate * 4) return@use null
                ShortArray(n) { input.readShort() }
            }
        }.getOrNull()
    }

    /**
     * Sessiz baş/sonu kırpar, tepe seviyesini eşitler ve uçlara kısa bir
     * yumuşatma koyar — ham kayıtta tık ve boşluk çok belirgin oluyor.
     */
    private fun clean(input: ShortArray): ShortArray? {
        val threshold = (Short.MAX_VALUE * 0.02).toInt()
        var start = 0
        while (start < input.size && abs(input[start].toInt()) < threshold) start++
        var end = input.size - 1
        while (end > start && abs(input[end].toInt()) < threshold) end--
        if (end <= start + 64) return null

        val out = input.copyOfRange(start, end + 1)
        val peak = out.maxOf { abs(it.toInt()) }
        if (peak > 8) {
            val gain = Short.MAX_VALUE * 0.85 / peak
            for (i in out.indices) {
                out[i] = (out[i] * gain).toInt().coerceIn(-32768, 32767).toShort()
            }
        }
        val ramp = min(256, out.size / 4)
        for (i in 0 until ramp) {
            val g = i.toDouble() / ramp
            out[i] = (out[i] * g).toInt().toShort()
            val j = out.size - 1 - i
            out[j] = (out[j] * g).toInt().toShort()
        }
        return out
    }

    private companion object {
        const val KEY_ENABLED = "customSoundsEnabled"
        /** Geri sayım adımı. iOS'takiyle aynı: 0,8 sn */
        const val TICK_MS = 800L
    }
}

File: Lumo/Services/CustomSoundStore.swift
import AVFoundation
import Combine

/// Premium oyuncunun kendi sesiyle değiştirebildiği efekt yuvaları.
/// Atlayış sesi oyunun ritmini taşıdığı için ayrı ele alınır: kayıt, kombo
/// yükseldikçe hızlandırılıp inceltilir — sentetik pentatonik dizinin yaptığını
/// oyuncunun kendi "hop"u yapar.
enum CustomSoundSlot: String, CaseIterable, Identifiable {
    case hop, collect, gate, lifeLost, fail, win

    var id: String { rawValue }

    /// Ayarlar ekranındaki başlık (yerelleştirme anahtarı)
    var title: String {
        switch self {
        case .hop:      return "Hop"
        case .collect:  return "Star"
        case .gate:     return "Gate"
        case .lifeLost: return "Life lost"
        case .fail:     return "Death"
        case .win:      return "Level finish"
        }
    }

    var hint: String {
        switch self {
        case .hop:      return "Rises in pitch as your combo grows."
        case .collect:  return "Plays when you pick up a star."
        case .gate:     return "Plays when the gate opens."
        case .lifeLost: return "Plays when an extra life is spent."
        case .fail:     return "Plays when the orb is lost."
        case .win:      return "Plays when a level is finished."
        }
    }

    var icon: String {
        switch self {
        case .hop:      return "arrow.up.forward"
        case .collect:  return "star.fill"
        case .gate:     return "lock.open.fill"
        case .lifeLost: return "heart.slash.fill"
        case .fail:     return "xmark.circle.fill"
        case .win:      return "flag.checkered"
        }
    }

    /// Kaydın azami süresi. "Hop" bir hece kadar olmalı; kombo hızlandığında
    /// uzun bir kayıt üst üste binip çamura dönüyor.
    var maxDuration: TimeInterval { (self == .hop || self == .collect) ? 0.9 : 2.0 }
}

/// Kayıtları diskte tutar, oyunun ses motoruna hazır tampon olarak verir.
/// Kayıtlar YALNIZCA cihazda kalır; hiçbir yere yüklenmez.
final class CustomSoundStore: NSObject, ObservableObject {
    static let shared = CustomSoundStore()

    /// Kaydı olan yuvalar
    @Published private(set) var recorded: Set<CustomSoundSlot> = []
    /// Şu an kayıt alınan yuva (nil = kayıt yok)
    @Published private(set) var recording: CustomSoundSlot?
    /// Geri sayımı süren yuva. Mikrofona basar basmaz kayıt başlayınca kimse
    /// yetişemiyordu; önce üçten geri sayıyoruz.
    @Published private(set) var arming: CustomSoundSlot?
    /// 3 → 2 → 1; kayıt başlayınca 0
    @Published private(set) var countdown = 0
    /// Kayıtlar oyunda kullanılsın mı (premium bitse bile kayıtlar silinmez)
    @Published var enabled: Bool {
        didSet {
            UserDefaults.standard.set(enabled, forKey: Self.enabledKey)
            applyToEngine()
        }
    }
    /// Mikrofon izni reddedildiyse ayarlarda uyarı gösterilir
    @Published private(set) var micDenied = false
    /// StoreManager'dan beslenir; premium bitince kayıtlar devre dışı kalır
    /// (dosyalar silinmez — abonelik dönerse yerinde durur).
    var premiumActive = false { didSet { if premiumActive != oldValue { applyToEngine() } } }

    private static let enabledKey = "customSoundsEnabled"
    private var recorder: AVAudioRecorder?
    private var stopWork: DispatchWorkItem?
    private var countdownWork: [DispatchWorkItem] = []
    /// Geri sayımın adım aralığı — bir metronom vuruşu kadar
    private let tick: TimeInterval = 0.8

    private let sampleRate: Double = 44_100
    private lazy var format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!

    private var folder: URL {
        let base = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let dir = base.appendingPathComponent("CustomSFX", isDirectory: true)
        if !FileManager.default.fileExists(atPath: dir.path) {
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir
    }

    private func url(for slot: CustomSoundSlot) -> URL {
        folder.appendingPathComponent("\(slot.rawValue).caf")
    }

    private override init() {
        let defaults = UserDefaults.standard
        // Varsayılan açık: kayıt yoksa zaten hiçbir şey değişmiyor
        enabled = defaults.object(forKey: Self.enabledKey) as? Bool ?? true
        super.init()
        refreshRecorded()
    }

    private func refreshRecorded() {
        recorded = Set(CustomSoundSlot.allCases.filter {
            FileManager.default.fileExists(atPath: url(for: $0).path)
        })
    }

    // MARK: Kayıt

    /// Mikrofon izni ister; verilirse üçten geri sayıp kaydı başlatır. Süre
    /// dolunca kendiliğinden durur — oyuncunun "dur" demeyi unutması bir yana,
    /// uzun kayıt zaten işe yaramıyor.
    func startRecording(_ slot: CustomSoundSlot) {
        guard recording == nil, arming == nil else { return }
        requestPermission { [weak self] granted in
            guard let self else { return }
            guard granted else {
                self.micDenied = true
                return
            }
            self.micDenied = false
            self.armRecording(slot)
        }
    }

    /// 3 → 2 → 1 → kayıt. Her adımda kısa bir tık çalar; ekrana bakmadan da
    /// ne zaman başlayacağı belli olsun diye.
    private func armRecording(_ slot: CustomSoundSlot) {
        arming = slot
        countdown = 3
        AudioEngine.shared.playTap()
        for step in 1...3 {
            let work = DispatchWorkItem { [weak self] in
                guard let self, self.arming == slot else { return }
                let left = 3 - step
                self.countdown = left
                if left > 0 {
                    AudioEngine.shared.playTap()
                } else {
                    self.arming = nil
                    self.beginRecording(slot)
                }
            }
            countdownWork.append(work)
            DispatchQueue.main.asyncAfter(deadline: .now() + tick * Double(step), execute: work)
        }
    }

    /// Geri sayarken vazgeçildi
    func cancelArming() {
        countdownWork.forEach { $0.cancel() }
        countdownWork = []
        arming = nil
        countdown = 0
    }

    private func requestPermission(_ done: @escaping (Bool) -> Void) {
        let handler: (Bool) -> Void = { granted in
            DispatchQueue.main.async { done(granted) }
        }
        if #available(iOS 17.0, *) {
            AVAudioApplication.requestRecordPermission(completionHandler: handler)
        } else {
            AVAudioSession.sharedInstance().requestRecordPermission(handler)
        }
    }

    private func beginRecording(_ slot: CustomSoundSlot) {
        countdownWork = []
        countdown = 0
        // Ambient pad mikrofona sızmasın diye motor kayıt boyunca duraklatılır
        AudioEngine.shared.stop()

        let session = AVAudioSession.sharedInstance()
        // Kayıt için oturumu geçici olarak playAndRecord'a alırız; biter bitmez
        // oyunun saygılı `.ambient` kategorisine geri dönülür.
        do {
            try session.setCategory(.playAndRecord, options: [.defaultToSpeaker, .mixWithOthers])
            try session.setActive(true)
        } catch {
            restoreSession()
            return
        }

        let settings: [String: Any] = [
            AVFormatIDKey: Int(kAudioFormatLinearPCM),
            AVSampleRateKey: sampleRate,
            AVNumberOfChannelsKey: 1,
            AVLinearPCMBitDepthKey: 32,
            AVLinearPCMIsFloatKey: true,
            AVLinearPCMIsBigEndianKey: false
        ]
        guard let rec = try? AVAudioRecorder(url: url(for: slot), settings: settings) else {
            restoreSession()
            return
        }
        recorder = rec
        recording = slot
        rec.record(forDuration: slot.maxDuration)

        // record(forDuration:) delegate'i her cihazda güvenilir tetiklemiyor;
        // kendi zamanlayıcımız durumu her hâlükârda temizler.
        let work = DispatchWorkItem { [weak self] in self?.finishRecording() }
        stopWork = work
        DispatchQueue.main.asyncAfter(deadline: .now() + slot.maxDuration + 0.15, execute: work)
    }

    /// Oyuncu "dur"a bastı — geri sayım sürüyorsa kayıt hiç başlamaz
    func stopRecording() {
        if arming != nil {
            cancelArming()
            return
        }
        guard recording != nil else { return }
        finishRecording()
    }

    private func finishRecording() {
        stopWork?.cancel()
        stopWork = nil
        recorder?.stop()
        recorder = nil
        recording = nil
        restoreSession()
        refreshRecorded()
        applyToEngine()
    }

    private func restoreSession() {
        let session = AVAudioSession.sharedInstance()
        try? session.setCategory(.ambient, options: [.mixWithOthers])
        try? session.setActive(true)
        AudioEngine.shared.resume()
    }

    func delete(_ slot: CustomSoundSlot) {
        try? FileManager.default.removeItem(at: url(for: slot))
        refreshRecorded()
        applyToEngine()
    }

    func hasRecording(_ slot: CustomSoundSlot) -> Bool { recorded.contains(slot) }

    /// Ayarlar ekranındaki "dinle" düğmesi
    func preview(_ slot: CustomSoundSlot) {
        guard let buffer = loadBuffer(slot) else { return }
        AudioEngine.shared.playPreview(buffer)
    }

    // MARK: Motora aktarma

    /// Kayıtları (premium ve açıksa) ses motoruna yükler. Premium biterse
    /// motordaki tamponlar boşaltılır ama dosyalar diskte kalır.
    func applyToEngine() {
        guard premiumActive, enabled else {
            AudioEngine.shared.clearCustomSounds()
            return
        }
        var set = CustomSoundSet()
        if let hop = loadBuffer(.hop) {
            set.hopLadder = AudioEngine.shared.pitchLadder(from: hop)
        }
        set.collect = loadBuffer(.collect)
        set.gate = loadBuffer(.gate)
        set.lifeLost = loadBuffer(.lifeLost)
        set.fail = loadBuffer(.fail)
        set.win = loadBuffer(.win)
        AudioEngine.shared.applyCustomSounds(set)
    }

    /// Dosyayı okur, sessiz baş/sonu kırpar, tepe seviyesini eşitler ve uçlara
    /// kısa bir yumuşatma koyar — ham kayıtta tık ve boşluk çok belirgin oluyor.
    private func loadBuffer(_ slot: CustomSoundSlot) -> AVAudioPCMBuffer? {
        let fileURL = url(for: slot)
        guard FileManager.default.fileExists(atPath: fileURL.path),
              let file = try? AVAudioFile(forReading: fileURL) else { return nil }
        let frames = AVAudioFrameCount(file.length)
        guard frames > 0,
              let raw = AVAudioPCMBuffer(pcmFormat: file.processingFormat, frameCapacity: frames),
              (try? file.read(into: raw)) != nil,
              let src = raw.floatChannelData?[0] else { return nil }

        let n = Int(raw.frameLength)
        var samples = [Float](repeating: 0, count: n)
        // Çok kanallı kaydı tek kanala indir (kayıt mono ama cihaz zorlayabiliyor)
        let channels = Int(raw.format.channelCount)
        if channels > 1, let data = raw.floatChannelData {
            for i in 0..<n {
                var sum: Float = 0
                for c in 0..<channels { sum += data[c][i] }
                samples[i] = sum / Float(channels)
            }
        } else {
            for i in 0..<n { samples[i] = src[i] }
        }

        // Sessizlik kırpma
        let threshold: Float = 0.02
        var start = 0
        while start < n, abs(samples[start]) < threshold { start += 1 }
        var end = n - 1
        while end > start, abs(samples[end]) < threshold { end -= 1 }
        guard end > start + 64 else { return nil }
        var trimmed = Array(samples[start...end])

        // Tepe eşitleme
        let peak = trimmed.reduce(Float(0)) { max($0, abs($1)) }
        if peak > 0.0001 {
            let gain = 0.85 / peak
            for i in trimmed.indices { trimmed[i] *= gain }
        }

        // Uçlara rampa (tık sesi olmasın)
        let ramp = min(256, trimmed.count / 4)
        for i in 0..<ramp {
            let g = Float(i) / Float(ramp)
            trimmed[i] *= g
            trimmed[trimmed.count - 1 - i] *= g
        }

        guard let out = AVAudioPCMBuffer(pcmFormat: format,
                                         frameCapacity: AVAudioFrameCount(trimmed.count)),
              let dst = out.floatChannelData?[0] else { return nil }
        out.frameLength = AVAudioFrameCount(trimmed.count)
        for i in trimmed.indices { dst[i] = trimmed[i] }
        return out
    }
}

/// Motora verilen hazır tamponlar
struct CustomSoundSet {
    /// Komboyla tizleşen atlayış sesleri (sentetik dizinin perde oranlarında)
    var hopLadder: [AVAudioPCMBuffer] = []
    var collect: AVAudioPCMBuffer?
    var gate: AVAudioPCMBuffer?
    var lifeLost: AVAudioPCMBuffer?
    var fail: AVAudioPCMBuffer?
    var win: AVAudioPCMBuffer?
}
